monash-human-power/raspicam
# Differentiate between null data and zero data on the data overlay

## Description

This PR definitely should not have taken a month to complete, but we here now so yeet. 🎉 

Having worked on this branch on and off during the last month, I am definitely sure that I missed something or there is a more cleaner way to write the code. So please do not hesitate to scrutinise the **hell** out of this PR. 

Merging in this PR should allow the Raspicam to recognise when data has stopped receiving after 5 seconds. When the Raspicam realises that data has not updated after 5 seconds, then the data overlay should display as `--` representing empty data. The reason behind this PR is for the rider to recognise 'no change' in sensor data on the overlay.

For example, if the rider is riding at a constant 50km/h for more than 5 seconds, then it should keep displaying 50km/h. But if the cadence was at 180 revolutions per minutes at one point, and the rider stops pedalling for more than 5 seconds, then it should update that there is no data here because the rider is not pedalling, whereas before it would keep showing 180 revolutions per minute.

## Screenshots

![image](https://user-images.githubusercontent.com/58575113/92999155-3996b680-f562-11ea-8e9d-a78f4dc14b96.png)
Peep my heavy eyebags due to very late nights.... Midsem break cannot come any quicker

## Steps to Test

1. Prepare three terminals, one in the Raspicam directory and the second in the data-acquisition-system/DAS/. Ensure there is a poetry virtual environment in both terminals. For the third terminal, it should run `mosquitto`.
2. In the DAS terminal, run `python das/tests/fake_module_tester.py` to load some fake data onto the mosquitto broker.
3. In the Raspicam terminal, run `python overlay_new.py` (or any of the overlays, but this is the most cleanest looking one imo)
4. Upon opening the Raspicam, there should be data updating onto the overlay. 
5. Stop the `fake_module_tester.py` in the DAS terminal, then quickly switch back onto the Raspicam overlay.
6. After some time (should be 5 seconds), the data points should disappear and appear "--" on the overlay.
